test/testRules.js

```diff
@@ -1183,6 +1184,74 @@ export const testRules = {
         });
         test.deepEqual(actual, expected);
 
+        test.done();
+    },
+
+    testResourceCompleteness: function(test) {
+        test.expect(5);
+
+        const rule = new ResourceCompleteness();
+        test.ok(rule);
+
+        const subjects = [
+            {},
+            { key: "completeness.test.source-missing", source: undefined },
+            { key: "completeness.test.target-missing", target: undefined },
+            { key: "completeness.test.both-missing", source: undefined, target: undefined },
+        ]
+            .map((overrides) => ({
+                key: "completeness.test",
+                sourceLocale: "en-US",
+                source: "Some source string.",
+                targetLocale: "de-DE",
+                target: "Some target string.",
+                pathName: "completeness-test.xliff",
+                state: "translated",
+                ...overrides, // override some of properties
+            }))
+            .map((props) => new ResourceString(props))
+            .map((resource) => ({ locale: "de-DE", resource, file: "x/y" }));
+
+        const results = subjects.map((subject) => rule.match(subject));
+
+        test.equal(results[0], undefined);
+
+        const commonResultProps = {
+            severity: "error",
+            rule,
+            pathName: "x/y",
+            locale: "de-DE",
+            description: "Resource must have both source and target element defined",
+        };
+        
+        test.deepEqual(
+            results[1],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.source-missing",
+                source: undefined,
+                highlight: "The following elements are missing in the resource: <e0>source</e0>",
+            })
+        );
+        test.deepEqual(
+            results[2],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.target-missing",
+                source: "Some source string.",
+                highlight: "The following elements are missing in the resource: <e0>target</e0>",
+            })
+        );
+        test.deepEqual(
+            results[3],
+            new Result({
+                ...commonResultProps,
+                id: "completeness.test.both-missing",
+                source: undefined,
+                highlight: "The following elements are missing in the resource: <e0>source, target</e0>",
```

Comment:
These should be 3 separate tests. The idea is that each unit test should test one condition only.

Reply:
Fixed